lightning/src/ln/functional_tests.rs

```diff
@@ -7402,22 +7414,60 @@ fn test_data_loss_protect() {
 	}
 
 	// Check we close channel detecting A is fallen-behind
+	// Check if we sent the warning message when we detecting that A is fallen-behind,
+	// and we give the possibility to A to  be able to recover from error.
 	nodes[1].node.handle_channel_reestablish(&nodes[0].node.get_our_node_id(), &reestablish_1[0]);
-	check_closed_event!(nodes[1], 1, ClosureReason::ProcessingError { err: "Peer attempted to reestablish channel with a very old local commitment transaction".to_string() });
-	assert_eq!(check_closed_broadcast!(nodes[1], true).unwrap().data, "Peer attempted to reestablish channel with a very old local commitment transaction");
-	check_added_monitors!(nodes[1], 1);
+	let warn_msg = "Peer attempted to reestablish channel with a very old local commitment transaction".to_owned();
+	assert!(check_warn_msg!(nodes[1], nodes[0].node.get_our_node_id(), chan.2).contains(&warn_msg));
 
 	// Check A is able to claim to_remote output
-	let node_txn = nodes[1].tx_broadcaster.txn_broadcasted.lock().unwrap().clone();
-	assert_eq!(node_txn.len(), 1);
-	check_spends!(node_txn[0], chan.3);
-	assert_eq!(node_txn[0].output.len(), 2);
-	mine_transaction(&nodes[0], &node_txn[0]);
-	connect_blocks(&nodes[0], ANTI_REORG_DELAY - 1);
-	check_closed_event!(nodes[0], 1, ClosureReason::ProcessingError { err: "We have fallen behind - we have received proof that if we broadcast remote is going to claim our funds - we can\'t do any automated broadcasting".to_string() });
-	let spend_txn = check_spendable_outputs!(nodes[0], node_cfgs[0].keys_manager);
-	assert_eq!(spend_txn.len(), 1);
-	check_spends!(spend_txn[0], node_txn[0]);
+	let mut node_txn = nodes[1].tx_broadcaster.txn_broadcasted.lock().unwrap().clone();
+	// The node B should not broadcast the transaction to force close the channel!
+	assert!(node_txn.is_empty());
+	// B should now detect that there is something wrong and should force close the channel.
+	let exp_err = "We have fallen behind - we have received proof that if we broadcast remote is going to claim our funds - we can\'t do any automated broadcasting";
+	check_closed_event!(nodes[0], 1, ClosureReason::ProcessingError { err: exp_err.to_string() });
+
+	// after the waring message sent by B, we should not able to
+	// use the channel, or reconnect with success to the channel.
+	match case {
+		DataLossProtectTestCase::UseChannel => {
+			assert!(nodes[0].node.list_usable_channels().is_empty());
```

Comment:
I'm a bit confused why having these two enum cases isn't identical coverage to moving this line right before the `peer_connected` calls below, and getting rid of the enum?

Reply:
The usage of the enum has a sense when we need to test different corner cases, where avoid using different boolean variables as a function parameter and make the code more readable!

Read this `do_test_data_loss_protect(DataLossProtectTestCase::TryToReconnect);` is better than this `do_test_data_loss_protect(true);` due the missing of name parameter in rust.

However, introducing an enum only for one case is not useful and you have right! I'm removing this. When I introduced it I was expecting that the branching was a little bit more complex!